core/src/main/scala/filodb.core/downsample/DownsampleCardinalityManager.scala

```diff
@@ -0,0 +1,170 @@
+        val numShards = conf.getInt("num-shards")
+        val result = numShards / minNumNodes
+        logger.info(s"Found the config to estimate the shards per node. minNumNodes=$minNumNodes numShards=$numShards" +
+          s" numShardsPerNode=$result")
+        result
+      case None =>
+        // NOTE: This is an Extremely UNLIKELY case, because the configs we rely on, are required for startup
+        // WHY 8? This is the current configuration of our production downsample cluster as of 06/02/2023
+        val defaultNumShardsPerNode = 8
+        logger.error(s"Could not find config for dataset=${rawDatasetRef.dataset} in configs " +
+          s"`dataset-configs` and `inline-dataset-configs`. Please check filodb config = ${filodbConfig.toString}" +
+          s" default-value=$defaultNumShardsPerNode")
+        defaultNumShardsPerNode
+    }
+  }
+
+  /**
+   * Determines if cardinality should be triggered at current hour of the day.
+   * For Example: With shardsPerNode = 8, the shard with number 0, should trigger cardinality when currentHour = 0,8,16
+   *
+   * NOTE: WHY are we using such criteria to trigger the cardinality count? -> This is done to stagger/distribute the
+   * calculation cost of cardinality count of different shards residing in the same node at different times of day.
+   * since cardinality count in downsample cluster is essentially a rebuild from scratch, which takes few mins. We don't
+   * want all the shards to be calculating this at the same time and using up all the resources of the node.
+   *
+   * @param currentShardNum Shard Number (0 - Num-Shards)
+   * @param shardsPerNode   Number of Shards per node
+   * @param currentHour     Current hour of the day (0-23)
+   * @return True if cardinality count is to be triggered, false otherwise
+   */
+  def shouldTriggerCardinalityCount(currentShardNum: Int, shardsPerNode: Int, currentHour: Int): Boolean = {
+    // check based on current time if this is the hour for current shard's cardinality count
+    require(shardsPerNode > 0)
+    (currentHour % shardsPerNode) == (currentShardNum % shardsPerNode)
+  }
+
+  /**
+   * Triggers cardinalityCount if metering is enabled and the required criteria matches.
+   * It creates a new instance of CardinalityTracker and uses the PartKeyLuceneIndex to calculate cardinality count
+   * and store data in a local CardinalityStore. We then close the previous instance of CardinalityTracker and switch
+   * it with the new one we created in this call.
+   */
+  def triggerCardinalityCount(): Unit = {
+    if (downsampleStoreConfig.meteringEnabled) {
+      try {
+        val currentHour = ((System.currentTimeMillis() / 1000 / 60 / 60) % 24).toInt
+        if (shouldTriggerCardinalityCount(shardNum, numShardsPerNode, currentHour)) {
+          // get a new cardinality tracker object and re-calculate the cardinality using this
+          val newCardTracker = getNewCardTracker()
```

Comment:
Should there be a tracker for each shard? Won't this setup only allow count data for at most one shard at any given time?

Reply:
We will have a tracker for each shard ( same for raw and DS cluster, since we are aiming to keep the setup similar to what we have in raw ( to minimize the query plan changes while fetching the data ).  

The idea here is that we don't want the physical node to be overwhelmed when all shards calculate cardinality at the same time. Hence, we are adding a check, which would only allow 1 shard to calculate its cardinality at a given point in time.